File: src/window.rs
use serde::Deserialize;
use tao::dpi::Size;
use tao::window::WindowAttributes;

use crate::webview::WebViewAttributesConfig;

#[derive(Deserialize, Debug, Default, Clone)]
pub struct WindowAttributesConfig {
    pub label: Option<String>,
    pub inner_size: Option<Size>,
    pub resizable: Option<bool>,
    pub minimizable: Option<bool>,
    pub maximizable: Option<bool>,
    pub closable: Option<bool>,
    pub title: Option<String>,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub maximized: Option<bool>,
    pub visible: Option<bool>,
    pub transparent: Option<bool>,
    pub decorations: Option<bool>,
    pub always_on_top: Option<bool>,
    pub always_on_bottom: Option<bool>,
    pub focused: Option<bool>,
    pub content_protection: Option<bool>,
    pub visible_on_all_workspaces: Option<bool>,
    pub background_color: Option<(u8, u8, u8, u8)>,
    pub webview: Option<WebViewAttributesConfig>,
}

impl From<WindowAttributesConfig> for WindowAttributes {
    fn from(config: WindowAttributesConfig) -> Self {
        WindowAttributes {
            inner_size: config.inner_size,
            resizable: config.resizable.unwrap_or(true),
            minimizable: config.minimizable.unwrap_or(true),
            maximizable: config.maximizable.unwrap_or(true),
            closable: config.closable.unwrap_or(true),
            title: config.title.unwrap_or_else(|| "PyWui window".to_string()),
            maximized: config.maximized.unwrap_or(false),
            visible: config.visible.unwrap_or(true),
            transparent: config.transparent.unwrap_or(false),
            decorations: config.decorations.unwrap_or(true),
            always_on_top: config.always_on_top.unwrap_or(false),
            always_on_bottom: config.always_on_bottom.unwrap_or(false),
            focused: config.focused.unwrap_or(true),
            content_protection: config.content_protection.unwrap_or(false),
            visible_on_all_workspaces: config.visible_on_all_workspaces.unwrap_or(false),
            background_color: config.background_color,
            ..WindowAttributes::default()
        }
    }
}

File: pywui_rs/window.py
from .pywui_rs import WindowManager

from .webview import Webview


class Window:
    def __init__(self, label: str, manager: WindowManager):
        self._manager = manager
        self._webview = Webview(label, manager)
        self._label = label

    def get_webview(self) -> Webview:
        return self._webview

    def resize(self, width: int, height: int):
        self._manager.update_window(
            self._label, {
                "width": width,
                "height": height
            })

    def fullscreen(self, full: bool):
        self._manager.update_window(
            self._label, {
                "fullscreen": full,
            })

    def show(self):
        self._manager.update_window(
            self._label, {
                "visible": True,
            })

    def hide(self):
        self._manager.update_window(
            self._label, {
                "visible": False,
            })

    def always_on_top(self, value: bool):
        self._manager.update_window(
            self._label, {
                "always_on_top": value,
            })

    def closable(self, value: bool):
        self._manager.update_window(
            self._label, {
                "closable": value,
            })

    def minimizable(self, value: bool):
        self._manager.update_window(
            self._label, {
                "minimizable": value,
            })

    def maximizable(self, value: bool):
        self._manager.update_window(
            self._label, {
                "maximizable": value,
            })

    def close(self):
        self._manager.close_window(self._label)

    def set_title(self, title: str):
        self._manager.update_window(
            self._label, {
                "title": title,
            })

    def background_color(self, r: int, g: int, b: int, a: int):
        self._manager.update_window(
            self._label, {
                "background_color": [r, g, b, a],
            })
